source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOTypeaheadDynamicChoicesService.mm

```diff
@@ -0,0 +1,107 @@
+    std::shared_ptr<ChoicesData> choicesData = choiceSet->GetChoicesData();
+    NSString *choicesDataType = [NSString stringWithCString:AdaptiveCardSchemaKeyToString(AdaptiveCardSchemaKey::ChoicesDataType).c_str() encoding:NSUTF8StringEncoding];
+    NSString *datasetKey = [NSString stringWithCString:AdaptiveCardSchemaKeyToString(AdaptiveCardSchemaKey::Dataset).c_str() encoding:NSUTF8StringEncoding];
+    NSString *value = [NSString stringWithCString:AdaptiveCardSchemaKeyToString(AdaptiveCardSchemaKey::Value).c_str() encoding:NSUTF8StringEncoding];
+    NSString *dataset = [NSString stringWithCString:choicesData->GetDataset().c_str()
+                                 encoding:NSUTF8StringEncoding];
+    NSString *dataQuery = [NSString stringWithCString:choicesData->GetChoicesDataType().c_str()
+                                 encoding:NSUTF8StringEncoding];
+
+    NSMutableDictionary *requestPayload = [NSMutableDictionary new];
+    [requestPayload setDictionary: @{
+        choicesDataType   : dataQuery,
+        datasetKey        : dataset,
+        value             : queryText ?: @"",
+        @"skip"           : skipVal, //used for pagination.Usage - retrieve top n records by skipping first x records.
+        @"top"            : topVal
+    }];
+    return requestPayload;
+}
+
+#pragma mark - ACOTypeaheadDebouncerDelegate Methods specifically for dynamic typeahead
+
+- (void)debouncerDidSendOutput:(id)key
+{
+    if ([key isKindOfClass:NSString.class])
+    {
+        __weak typeof(self) weakSelf = self;
+        dispatch_async(_global_queue, ^{
+            __strong typeof(self) strongSelf = weakSelf;
+            if ([strongSelf->_rootView.acrActionDelegate respondsToSelector:@selector(onChoiceSetQueryChange:acoElem:completion:)]) {
+                NSDictionary *requestPayload = [strongSelf getRequestPayloadForTypeaheadSearchWithQueryText:key withSkipValue:[NSNumber numberWithInt:0] AndTopValue:[NSNumber numberWithInt:15]];
+                [strongSelf->_rootView.acrActionDelegate onChoiceSetQueryChange:requestPayload acoElem:strongSelf->_inputElem completion:^(NSDictionary * response, NSError *error) {
+                    dispatch_async(dispatch_get_main_queue(), ^{
+                        NSDictionary *choices = [response objectForKey:@"value"];
+                        if ([strongSelf->_typeaheadSearchDelegate respondsToSelector:@selector(updateTypeaheadUIWithSearchText:dynamicChoices:withError:)]) {
+                            [strongSelf->_typeaheadSearchDelegate updateTypeaheadUIWithSearchText:key dynamicChoices:choices withError:error];
+                        }
+                    });
+                }];
+            } else {
+                if ([strongSelf->_typeaheadSearchDelegate respondsToSelector:@selector(updateTypeaheadUIWithSearchText:dynamicChoices:withError:)]) {
+                    [strongSelf->_typeaheadSearchDelegate updateTypeaheadUIWithSearchText:key dynamicChoices:nil withError:nil];
```

Comment:
wouldn't it be better to have dedicated error handler instead of overloading updateTypeaheadUIWithSearchText to update ui for error.

Reply:
updated